qiskit/circuit/bit.py

```diff
@@ -120,3 +121,20 @@ def __eq__(self, other):
             return self._repr == other._repr
         except AttributeError:
             return False
+
+    def __copy__(self):
+        # Bits are immutable.
+        return self
+
+    def __deepcopy__(self, memo=None):
+        if (self._register, self._index) == (None, None):
+            return self
+
+        # Old-style bits need special handling for now, since some code seems
+        # to rely on their registers getting deep-copied.
+        bit = type(self).__new__(type(self))
+        bit._register = copy.deepcopy(self._register, memo)
+        bit._index = self._index
+        bit._hash = self._hash
+        bit._repr = self._repr
+        return bit
```

Comment:
Yeah, in fairness, we were wrong before when we didn't recurse the `deepcopy` onto the register.  It likely would have caused a weird inconsistency where the `reg[0]._register is not reg`, which is an invariant of old-style registers/bits at the moment.  I can't immediately see _why_ that has knock-on effects, but I guess it shouldn't be surprising.